araig_calculators/src/loggers/rosbagger_class.py

```diff
@@ -3,84 +3,73 @@
+            self.node_name + "/whitelist",
+            "/test_type"]
 
+        super(RosbaggerClass, self).__init__(param_list = param_list)
+
+        try:
+            while not rospy.is_shutdown():
+                self.main_loop() 
+        except rospy.ROSException:
+            pass 
+    
+    def prepare_topics(self):
         list_of_topics = rospy.get_published_topics()
         list_of_topics.sort()
-
-        self._topics_string = ""
+        topics_string = ""
         for topic, msg in list_of_topics:
-            if "/image_raw" in topic:
-                if (not "/image_raw/compressed" in topic) or ("compressedDepth" in topic) :
-                    rospy.loginfo(rospy.get_name() + ": Ignoring topic - {}".format(topic))
-                    continue
-                else:
-                    self._topics_string = self._topics_string + topic + " "
-            else:
-                self._topics_string = self._topics_string + topic + " "
-
-        # Prepare command
-        self._command = "rosbag record -o " + self.filename_ + " " + self._topics_string
-        # rospy.loginfo(rospy.get_name() + ": Ready to execute: " + self._command)
-        self.main_loop()
-        rospy.spin()
-
-    def __del__(self):
-        rospy.loginfo(rospy.get_name() + ": Destructor killing process {}".format(self._rosbag_proc))
-        self._rosbag_proc.send_signal(subprocess.signal.SIGINT)
-
-    def stop_callback(self, msg):
-            with self.stop_mutex_:
-                self._stop_recording_ = msg.data
-
-    def start_callback(self, msg):
-            with self.start_mutex_:
-                self._start_recording = msg.data
+            for black_topic in self.config_param[self.node_name + "/blacklist"]:
+                if topic == black_topic:
+                    topic = ""
+            for white_topic in self.config_param[self.node_name + "/whitelist"]:
+                if topic == white_topic:
+                    topic = ""
+            topics_string = topics_string + topic + " "
+        
+        for white_topic in self.config_param[self.node_name + "/whitelist"]:
+            topics_string = topics_string + white_topic + " "
+        
+        return topics_string
 
     def main_loop(self):
         rospy.loginfo(rospy.get_name() + ": Waiting for start signal...")
         # Wait for start signal
         start = False
+        stop = False
         while not start and not rospy.is_shutdown():
             self._rate.sleep()
-            with self.stop_mutex_:
-                start = self._start_recording
-        
-        rospy.loginfo(rospy.get_name() + ": Starting recording now!")
-            
-        command = shlex.split(self._command)
-        self._rosbag_proc = subprocess.Popen(command)
+            start = self.getSafeFlag("start")
+    
+         # get published topic
+        if start == True and stop == False:
+            rospy.sleep(self.config_param[self.node_name + "/start_offset"])
```

Comment:
Do we really need a start offset sleep here? The runner already has a sleep which is meant for the rosbagger to spool up, having these sleep here as well is actually counterproductive.

Reply:
Okay I see what you have done here. You start the rosbagger after a start offset sleep after ui start signal is received. But this does not solve our problem, because the rosbagger starts recording after the sleep, whereas we need to start recording before the sleep, in order to let it get started properly. Therefore, this sleep needs to be in the runner not here. (cant start recording before we get start signal :wink:  )

Reply:
But then every logger seems to have a start offset. Unclear. Need to discuss. From my perspective:   
User clicks start -> All loggers start logging -> sleep(2s) -> Give start command to robot   
This gives enough time for loggers to spool up. But it seems like in all these cases, loggers start after the sleep.

Reply:
bagger TODO:KUT    
other TODO:KUT